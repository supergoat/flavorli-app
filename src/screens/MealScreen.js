// @flow
import React, { Component } from "react";
import { SafeAreaView, Text, View } from "react-native";
import FAColors from "../common/FAColors";
import FAStyleSheet from "../common/FAStyleSheet";
import FAButton from "../common/FAButton";
import FACircleButton from "../common/FACircleButton";
import type { NavigationScreenProp } from "react-navigation";
import type { MealType } from "../common/FATypes";

/** ============================================================================
<MealScreen
  meal={MealType}
  navigation={NavigationScreenProp}
/>
--------------------------------------------------------------------------------
Displays the meal, and provides and interface for the viewer to add it to the
basket
============================================================================= */
type State = {
  quantity: number
};
type Props = {
  navigation: NavigationScreenProp
};
class MealScreen extends Component<Props, State> {
  state = { quantity: 1 };

  render() {
    const { navigation } = this.props;
    const { quantity } = this.state;

    const meal = navigation.getParam("meal", null);

    return (
      <SafeAreaView style={{ flex: 1, backgroundColor: FAColors.alabaster }}>
        <View style={styles.container}>
          <Text style={styles.mealName}>
            {meal.name}
          </Text>

          <Text style={styles.mealPrice}>
            £{meal.price.toFixed(2)}
          </Text>

          <Text style={styles.mealDescription}>
            {meal.description}
          </Text>

          <View style={styles.buttons}>
            <FACircleButton
              disabled={quantity <= 1}
              onPress={this.decreaseQty}
              text="-"
            />

            <Text style={styles.qtyText}>{quantity}</Text>

            <FACircleButton
              disabled={quantity >= 10}
              onPress={this.increaseQty}
              text="+"
            />
          </View>

          <View style={styles.buttons}>
            <FAButton
              type="secondary"
              onPress={() => navigation.goBack()}
              style={styles.cancelButton}
              text="Cancel"
            />

            <FAButton
              type="primary"
              onPress={() => navigation.goBack()}
              style={styles.addButton}
              text={`Add For £${(meal.price * quantity).toFixed(2)}`}
            />
          </View>
        </View>
      </SafeAreaView>
    );
  }

  /**
   * Increase the quantity by 1
   */
  increaseQty = () => {
    const { quantity } = this.state;
    if (quantity < 10) this.setState({ quantity: quantity + 1 });
  }

  /**
   * Decreases the quantity by 1
   */
  decreaseQty = () => {
    const { quantity } = this.state;
    if (quantity > 1) this.setState({ quantity: quantity - 1 });
  }
}

/* StyleSheet
============================================================================= */
const styles = FAStyleSheet.create({
  container: {
    flex: 1,
    padding: 15,
    backgroundColor: FAColors.alabaster
  },
  mealName: {
    fontFamily: "Lato",
    fontSize: 30,
    color: FAColors.oxfordBlue,
    paddingVertical: 10
  },
  mealPrice: {
    fontFamily: "Lato",
    fontSize: 20,
    color: FAColors.rollingStone
  },
  mealDescription: {
    fontFamily: "Lato",
    paddingTop: 20,
    fontSize: 18,
    color: FAColors.osloGrey
  },
  buttons: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    paddingTop: 40
  },
  qtyText: {
    fontSize: 25,
    fontFamily: "Lato",
    fontWeight: "bold",
    marginHorizontal: 20,
    width: 40,
    textAlign: "center"
  },
  cancelButton: {
    marginRight: 15
  },
  addButton: {
    flex: 2
  }
});

/* Export
============================================================================= */
export default MealScreen;
