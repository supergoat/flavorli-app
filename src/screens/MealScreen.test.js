// @flow
import React from 'react';
import { View, Text } from 'react-native';
import MealScreen from './MealScreen';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

const meal = {
  id: '1',
  name: 'Lasagna',
  description: 'This is the lasagna description',
  price: 12.00
}

describe('MealScreen', () => {
  it('renders correctly', () => {
    const getParamMock = jest.fn(() => meal)
    const tree = renderer.create(<MealScreen navigation={{ getParam: getParamMock }} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('does not decrease the quantity if its equal to 1', () => {
    const getParamMock = jest.fn(() => meal)
    const component = shallow(<MealScreen navigation={{ getParam: getParamMock }} />)

    expect(component.state('quantity')).toEqual(1)
    // Press minus button
    component.find('FACircleButton').at(0).prop('onPress')();
    expect(component.state('quantity')).toEqual(1)
  });

  it('decreases the quantity by 1 if greater than 1', () => {
    const getParamMock = jest.fn(() => meal)
    const component = shallow(<MealScreen navigation={{ getParam: getParamMock }} />)

    // Set the state to 2
    component.setState({ quantity: 2})

    expect(component.state('quantity')).toEqual(2)
    // Press plus button
    component.find('FACircleButton').at(0).prop('onPress')();
    expect(component.state('quantity')).toEqual(1)
  });

  it('does not increases the quantity by 1 if its equal to 10', () => {
    const getParamMock = jest.fn(() => meal)
    const component = shallow(<MealScreen navigation={{ getParam: getParamMock }} />)

    // Set the state to 10
    component.setState({ quantity: 10})

    expect(component.state('quantity')).toEqual(10)
    // Press minus button
    component.find('FACircleButton').at(1).prop('onPress')();
    expect(component.state('quantity')).toEqual(10)
  });

  it('increases the quantity by 1 if its less than 10', () => {
    const getParamMock = jest.fn(() => meal)
    const component = shallow(<MealScreen navigation={{ getParam: getParamMock }} />)

    expect(component.state('quantity')).toEqual(1)
    // Press plus button
    component.find('FACircleButton').at(1).prop('onPress')();
    expect(component.state('quantity')).toEqual(2)
  });

  it('calls goBack when cancel button is pressed', () => {
    const getParamMock = jest.fn(() => meal)
    const goBackMock = jest.fn(() => meal)
    const component = shallow(
      <MealScreen
        navigation={{ getParam: getParamMock, goBack: goBackMock }}
      />
    );

    // Press cancel button
    component.find('FAButton').at(0).prop('onPress')();
    expect(goBackMock).toHaveBeenCalled();
  });

  it('calls goBack when add button is pressed', () => {
    const getParamMock = jest.fn(() => meal)
    const goBackMock = jest.fn(() => meal)
    const component = shallow(
      <MealScreen
        navigation={{ getParam: getParamMock, goBack: goBackMock }}
      />
    );

    // Press add button
    component.find('FAButton').at(1).prop('onPress')();
    expect(goBackMock).toHaveBeenCalled();
  });
});
