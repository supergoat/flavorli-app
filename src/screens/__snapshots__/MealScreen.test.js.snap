// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealScreen renders correctly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "#F9F9F9",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#F9F9F9",
        "flex": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#333C4E",
          "fontFamily": "Lato",
          "fontSize": 30,
          "paddingVertical": 10,
        }
      }
    >
      Lasagna
    </Text>
    <Text
      style={
        Object {
          "color": "#788087",
          "fontFamily": "Lato",
          "fontSize": 20,
        }
      }
    >
      £
      12.00
    </Text>
    <Text
      style={
        Object {
          "color": "#7e8890",
          "fontFamily": "Lato",
          "fontSize": 18,
          "paddingTop": 20,
        }
      }
    >
      This is the lasagna description
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingTop": 40,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderColor": "#333C4E",
            "borderRadius": 19,
            "borderWidth": 1,
            "height": 38,
            "justifyContent": "center",
            "opacity": 0.5,
            "width": 38,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#333C4E",
              "fontFamily": "Lato",
              "fontSize": 25,
              "fontWeight": "bold",
            }
          }
        >
          -
        </Text>
      </View>
      <Text
        style={
          Object {
            "fontFamily": "Lato",
            "fontSize": 25,
            "fontWeight": "bold",
            "marginHorizontal": 20,
            "textAlign": "center",
            "width": 40,
          }
        }
      >
        1
      </Text>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderColor": "#333C4E",
            "borderRadius": 19,
            "borderWidth": 1,
            "height": 38,
            "justifyContent": "center",
            "opacity": 1,
            "width": 38,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#333C4E",
              "fontFamily": "Lato",
              "fontSize": 25,
              "fontWeight": "bold",
            }
          }
        >
          +
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingTop": 40,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderColor": "#333C4E",
            "borderRadius": 3,
            "borderWidth": 1,
            "flex": 1,
            "marginRight": 15,
            "opacity": 1,
            "padding": 15,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Lato",
              },
              Object {
                "color": "#333C4E",
              },
            ]
          }
        >
          Cancel
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#333C4E",
            "borderRadius": 3,
            "borderWidth": 1,
            "flex": 2,
            "opacity": 1,
            "padding": 15,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Lato",
              },
              false,
            ]
          }
        >
          Add For £12.00
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
